src/irmin-pack/unix/lower.ml

```diff
@@ -0,0 +1,195 @@
+
+  let set_readonly ~readonly t =
+    if t.readonly = readonly then Ok ()
+    else
+      (* TODO actually reopen based on readonly flag once
+         sparse file is in place *)
+      Ok (t.readonly <- readonly)
+
+  let create_empty volume_path =
+    let open Result_syntax in
+    (* 0. Validate volume directory does not already exist *)
+    let* () =
+      match Io.classify_path volume_path with
+      | `Directory | `File | `Other -> Error (`File_exists volume_path)
+      | `No_such_file_or_directory -> Ok ()
+    in
+    (* 1. Make directory *)
+    let* () = Io.mkdir volume_path in
+    (* 2. Make empty mapping *)
+    let* () =
+      Irmin_pack.Layout.V5.Volume.(
+        Io.create ~path:(mapping ~root:volume_path) ~overwrite:true)
+      >>= Io.close
+    in
+    (* 3. Make empty data *)
+    let* () =
+      Irmin_pack.Layout.V5.Volume.(
+        Io.create ~path:(data ~root:volume_path) ~overwrite:true)
+      >>= Io.close
+    in
+    (* TODO: handle failure to create all artifacts, either here or in a cleanup
+       when the store starts. *)
+    v ~readonly:false volume_path
+
+  let path t = t.path
+  let control t = t.control
+  let empty_state t = if Option.is_none t.control then `Empty t else `Nonempty t
+
+  let is_empty t =
+    match empty_state t with `Empty _ -> true | `Nonempty _ -> false
+
+  let nonempty_control (`Nonempty t) =
+    match t.control with
+    | None -> failwith "Nonempty volume has no control"
+    | Some cf -> cf
+
+  let start_offset t = (nonempty_control t).start_offset
+  let end_offset t = (nonempty_control t).end_offset
+
+  let compare a b =
+    match (empty_state a, empty_state b) with
+    | `Empty _, `Empty _ -> 0
+    (* Always sort empty volumes last. *)
+    | `Empty _, `Nonempty _ -> 1
+    | `Nonempty _, `Empty _ -> -1
+    | (`Nonempty _ as a), (`Nonempty _ as b) ->
+        Int63.compare (start_offset a) (start_offset b)
+
+  let contains ~offset t =
+    match empty_state t with
+    | `Empty _ -> false
+    | `Nonempty _ as t ->
+        let open Int63.Syntax in
+        start_offset t <= offset && offset < end_offset t
+end
+
+module Make (Io : Io.S) (Errs : Io_errors.S with module Io = Io) = struct
+  module Io = Io
+  module Errs = Errs
+  module Volume = Make_volume (Io) (Errs)
+
+  type t = { root : string; readonly : bool; mutable volumes : Volume.t array }
+  type open_error = [ Volume.open_error | `Not_a_directory of string ]
+  type close_error = [ | Io.close_error ]
+
+  type add_error =
+    [ open_error
+    | `Ro_not_allowed
+    | `Multiple_empty_volumes
+    | `File_exists of string
+    | `Invalid_parent_directory ]
+
+  exception LoadVolumeError of Volume.open_error
+
+  let load_volumes t =
+    let open Result_syntax in
+    let root = t.root in
+    let* volume_paths =
+      match Io.classify_path root with
+      | `Directory ->
+          Ok (Sys.readdir root |> Array.map (fun p -> Filename.concat root p))
```

Comment:
I think the expected number of volumes should be obtained before hand from the root control file and that the filenames should be inferred from there (just like chunks), rather than rely on `readdir` (its ordering is likely wrong -- and we don't want a RO node reloading and seeing a partial volume while we are in the process of creating it)

Reply:
I agree with you. It's an idea that we discussed a couple of weeks ago, but for some reason, I had in my mind that the control file changes needed to come before this PR (and I wanted to work on this code first), but it's not needed since this PR doesn't yet hook into store opening process. I'll make this change!

(also, yeah RO calculation is totally wrong since the sort happens after volume creation :see_no_evil:)